### PR TITLE
Simpler/smaller/more performant GadgetRecord

### PR DESCRIPTION
**Pull request by airhorns**
While poking at the `GadgetRecord` class for some other stuff, I noticed some inefficiency in how the change tracking is implemented. `GadgetRecord` is a strange object because it is a class that has all our business logic wrapped in a proxy that allows us to do the change tracking. The objects we had back into userland are `Proxy` objects that intercept all `get` and `set` calls to do field change tracking. The actual `GadgetRecord` object sits "next" to the proxy, in that it isn't even the proxy target, but instead an object that is closed over by the proxy to provide some of the functionality. 

Previously, `GadgetRecord` stuck all its private state on a `__gadget` object on it. This adds one more object allocation to every created record object and an unnecessary layer of indirection. the `GadgetRecord` object itself is already an object with a different identity than the place where we store all the actual keys/values of the record. I think `__gadget` was from before that was the case. So, this removes it in favour of just storing the bits right on the `GadgetRecord` itself. To keep the proxy logic simple for what is a field and what is a special key, I stored all the bits under symbols, which still makes it pretty clear that it is special.

I also removed one of the special bits as I think it was redundant, we had a list of the keys we were tracking in an array, as well as a map of {[key]: true}, both always set at the same time.

### CODE REVIEW

File: packages/api-client-core/src/GadgetRecord.ts [hunk 1]
```diff
@@ -9,66 +9,75 @@ export enum ChangeTracking {
...
+  [kTouched] = false;
 
   private empty = false;
 
   constructor(data: Shape) {
-    this.__gadget.instantiatedFields = cloneDeep(data);
-    this.__gadget.persistedFields = cloneDeep(data);
-    Object.assign(this.__gadget.fields, data);
+    this[kInstantiatedFields] = cloneDeep(data) ?? {};
+    this[kPersistedFields] = cloneDeep(data) ?? {};
```
Comment (airhorns) [inline packages/api-client-core/src/GadgetRecord.ts:34]: I tried making this lazy but it breaks our contracts around detecting nested changes, so we still gotta do this eagerly on instantiation :(

File: packages/api-client-core/src/GadgetRecord.ts [hunk 2]
```diff
@@ -9,66 +9,75 @@ export enum ChangeTracking {
 
 export type RecordShape = Record<string, any> | null | undefined | void;
 
+const kFields = Symbol.for("g/fields");
+const kInstantiatedFields = Symbol.for("g/if");
+const kPersistedFields = Symbol.for("g/pf");
+const kFieldKeys = Symbol.for("g/fk");
+const kTouched = Symbol.for("g/t");
```
Comment (thegedge) [inline packages/api-client-core/src/GadgetRecord.ts:16]: nit: Any reason to not go for clearer names?

File: packages/api-client-core/src/GadgetRecord.ts [hunk 3]
```diff
@@ -9,66 +9,75 @@ export enum ChangeTracking {
...
     }
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
 
     const handler = {
       get: (obj: any, prop: string | symbol) => {
-        if (prop in this) {
-          // if the GadgetRecordImplementation responds to the property or function
-          // prioritize that
-          return (this as any)[prop];
```
Comment (thegedge) [inline packages/api-client-core/src/GadgetRecord.ts:42]: `this === handler` here, eh? So it wasn't doing anything useful :o

**Review by thegedge — APPROVED** [review]
🚀

File: packages/api-client-core/src/GadgetRecord.ts (hunk 2 above)
Comment (airhorns) [reply packages/api-client-core/src/GadgetRecord.ts:16]: Yeah, bundle size -- these strings don't get mangled during minification. EVERY BYTE COUNTS

File: packages/api-client-core/src/GadgetRecord.ts (hunk 1 above)
Comment (airhorns) [reply packages/api-client-core/src/GadgetRecord.ts:34]: Yup, or just only doing the copy the first time we need to write something. But, we have tests that say "modifying an array property marks that property as changed", which means we need to either proxy nested stuff in order to catch the write to the nested doodad, or just keep eagerly deepcloning. It... sucks.

**Comment by airhorns** [comment]
Gadget CI passes with this here: https://buildkite.com/gadget/gadget/builds/23302#018fc0c4-ad50-47cc-9398-d80ffb0eda50

**Pull request merged**